### PR TITLE
Allow structured data types in Distribution

### PR DESCRIPTION
**Pull request by mhvk**
### Description

This pull request is to allow structured data types for `Distribution`, so that classes like `EarthLocation` can be made into `Distribution` as well (though that will need some changes in `EarthLocation` itself).

To make this possible, an API change was necessary, where the `.dtype` exposed by the distribution now equals the `.dtype` of the samples (rather than the structured dtype used internally). I could not get `QuantityDistribution` to work without changes to `Quantity` and `StructuredUnit` that I felt were not warranted. And arguably, it is more logical for the distribution to look as much as possible like a regular array, so to expose the regular `dtype`.

Note that this builds on #15295; it may be best to get that in first. Otherwise, reviewing by commit may be helpful.

- [x] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

File: astropy/uncertainty/core.py [hunk 1]
```diff
@@ -43,40 +46,142 @@ def __new__(cls, samples):
...
-            # arrays are typeset, leading to all samples to be shown.  It may
-            # be possible to hack oneself out by temporarily becoming a void.
-            new_name = samples_cls.__name__ + cls.__name__
+            # Walk through MRO and find closest base data class.
+            # Note: right now, will basically always be ndarray, but
+            # one could imagine needing some special care for one subclass,
+            # which would then get its own entry.  E.g., if MaskedAngle
+            # defined something special, then MaskedLongitude should depend
+            # on it.
+            for mro_item in samples_cls.__mro__:
```
Comment (nstarman) [inline astropy/uncertainty/core.py:113]: Nitpick. I think 
```suggestion
            for mro_item in samples_cls.mro():
```
is preferred.

File: docs/changes/uncertainty/15296.api.rst [hunk 2]
```diff
@@ -0,0 +1,16 @@
+The ``.dtype`` attribute exposed by ``Distribution`` is now that of
+the samples, rather than one that has a "samples" entry.  This makes
+quantities with structured data types and units easier to support, and
+generally makes the ``Distribution`` appear more simular to regular
```
Comment (nstarman) [inline docs/changes/uncertainty/15296.api.rst:4]:
```suggestion
generally makes the ``Distribution`` appear more similar to regular
```

**Review by nstarman — COMMENTED** [review]
There's some deep numpy foo, so this is not a full review. I just glanced through. From the tests, everything looks good

File: astropy/uncertainty/core.py (hunk 1 above)
Comment (mhvk) [reply astropy/uncertainty/core.py:113]: I didn't know this, but https://stackoverflow.com/questions/19713585/whats-the-difference-between-the-mro-method-and-the-mro-attribute-of-a-clas suggests that `.mro()` is called at class creation time and caches it results in `__mro__`. If so, I happened to have done the right thing...

File: docs/changes/uncertainty/15296.api.rst (hunk 2 above)
Comment (mhvk) [reply docs/changes/uncertainty/15296.api.rst]: Done

File: astropy/uncertainty/core.py [hunk 3]
```diff
@@ -34,7 +35,9 @@ class Distribution:
     samples : array-like
         The distribution, with sampling along the *leading* axis. If 1D, the
         sole dimension is used as the sampling axis (i.e., it is a scalar
-        distribution).
+        distribution).  If an `~numpy.ndarray` or subclass, the data will
```
[the remark this answers is not in the document]
Comment (mhvk) [reply astropy/uncertainty/core.py]: Ah, yes `|ndarray|` it is!

**Comment by mhvk** [comment]
#15304, which includes this, was reviewed as a whole and will get merged, so am closing this one.

**Pull request closed without merging**